Datadog SDK OTel Metrics and Logs API Support: Phase 1

brett0000FF opened the pull request:
<!-- *Note: Please remember to review the Datadog Documentation [Contribution Guidelines](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md) if you have not yet done so.* -->

### What does this PR do? What is the motivation?
<!-- A brief description of the change being made with this pull request. What is your motivation for the PR? -->

Document phase 1 of OTel Metrics and Logs API support for:

- Node.js
- Python
- .NET

#### Previews
- [OpenTelemetry Environment Variables Interoperability](https://docs-staging.datadoghq.com/brett.blue/otel-log-metric-phase-1/opentelemetry/config/environment_variable_support) (Updated)
- [OpenTelemetry API Support](https://docs-staging.datadoghq.com/brett.blue/otel-log-metric-phase-1/opentelemetry/instrument/api_support/) (Updated)
- [Telemetry/Language-Specific Pages](https://docs-staging.datadoghq.com/brett.blue/otel-log-metric-phase-1/opentelemetry/instrument/api_support/dotnet/traces) (New Metrics and Logs pages)

#### Changes
- Updated `content/en/opentelemetry/instrument/api_support/_index.md` to include metrics and logs details.
- Created reusable shortcodes to use across pages to make maintenance easier.
   - `layouts/shortcodes/otel-api-troubleshooting.en.md`: Common troubleshooting content.
   - `layouts/shortcodes/otel-overview-exporter.en.md` : Exporter-based setup overview content.
   - `layouts/shortcodes/otel-overview-native.en.md`: Native-based setup overview content.
- Added new environment variables and log/metrics content to `content/en/opentelemetry/config/environment_variable_support.md`.
- Added page-level tabs for Traces, Metrics, and Logs for Python, Node.js, and .NET.
- Added new API support table to `/opentelemetry/compatibility/`.

### Merge instructions

<!-- 
If you're waiting for a release or there are other considerations that you want us to be aware of, list them here. 
If the PR is ready to be merged once it receives the required reviews, check the box below after you've created the PR.
-->

Merge readiness:
- [ ] Ready for merge

**For Datadog employees**:

Your branch name MUST follow the `<name>/<description>` convention and include the forward slash (`/`). Without this format, your pull request will not pass CI, the GitLab pipeline will not run, and you won't get a branch preview. Getting a branch preview makes it easier for us to check any issues with your PR, such as broken links.

If your branch doesn't follow this format, rename it or create a new branch and PR.

[6/5/2025] Merge queue has been disabled on the documentation repo. If you have write access to the repo, the PR has been reviewed by a Documentation team member, and all of the required checks have passed, you can use the **Squash and Merge** button to merge the PR. If you don't have write access, or you need help, reach out in the #documentation channel in Slack.

### Additional notes
<!-- Anything else we should know when reviewing?-->

<!-- Previewing the PR: Assuming you are a Datadog employee and named your branch `<yourname>/<description>`, a preview build will run and links to the preview output will be auto-generated and posted in the PR comments. The links will 404 until the preview build is finished running. -->

mabdinur commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -97,6 +108,46 @@ Disable the SDK for all signals<br>
+`OTEL_EXPORTER_OTLP_ENDPOINT`
+: **Description**: Specifies the URL for sending OTLP data. <br>
+**Default (gRPC)**: `http://localhost:4317`. <br>
+**Default (HTTP)**: `http://localhost:4318/v1/metrics`.
```
`/v1/metrics` is metric specific

```suggestion
**Default (HTTP)**: `http://localhost:4318`.
```

mabdinur commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -97,6 +108,46 @@ Disable the SDK for all signals<br>
+Datadog SDKs support the following OpenTelemetry Metrics SDK options.
+
+`OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE`
+: **Description**: Specifies the aggregation `temporality` to use for each instrument kind. <br>
```
We can also link to this page for more details: https://opentelemetry.io/docs/specs/otel/metrics/sdk_exporters/otlp/#additional-environment-variable-configuration

mabdinur commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,96 @@
+`OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`
+: **Description**: Specifies the URL for sending OTLP logs. Takes precedence over `OTEL_EXPORTER_OTLP_ENDPOINT`.
+: **Default (gRPC)**: `http://localhost:4317`
+: **Default (HTTP)**: `http://localhost:4318/v1/logs`
```
These configurations will be used in all otel log implementations. Should we move the definitions to the content/en/opentelemetry/config/environment_variable_support.md page?

mabdinur commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
````diff
@@ -0,0 +1,102 @@
+
+1. Install the Datadog SDK:
+   ```sh
+   [Code example]
````
@link04 can you fill this in

mabdinur commented on content/en/opentelemetry/instrument/api_support/nodejs/metrics.md:
```diff
@@ -0,0 +1,102 @@
+
+### Existing DogStatsD setup
+
+If you are currently using the Datadog DogStatsD client and want to migrate to the OpenTelemetry Metrics API, you need to update your instrumentation code. The main difference is that OTel metrics are configured using environment variables rather than code, and you create `Instrument` objects first.
```
We can link to opentelemetry nodejs onboarding docs here: https://github.com/open-telemetry/opentelemetry-js/blob/4dded45ec67f90998d75e3ddbde9872b6b466183/doc/metrics.md#getting-started

duncanhewett commented on content/en/opentelemetry/instrument/api_support/python/metrics.md:
```diff
@@ -0,0 +1,110 @@
+## Prerequisites
+
+- **Datadog SDK**: dd-trace-py version 3.18.0 or later.
+- **OpenTelemetry SDK**: The standard opentelemetry-sdk package for Python.
```
We have steps in the setup for installing the SDK and Exporter, so it could be confusing to have them as prerequisites as well?

duncanhewett commented on content/en/opentelemetry/instrument/api_support/python/logs.md:
````diff
@@ -0,0 +1,131 @@
+
+This example shows how logs emitted within an active Datadog span are automatically correlated.
+
+```python
````
(Example needs completing)

zacharycmontoya commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -71,38 +86,153 @@ Mapped values between `OTEL_TRACES_SAMPLER` & `DD_TRACE_SAMPLE_RATE`:<br>
+: **Description**: Specifies the aggregation `temporality` to use for each instrument kind. <br>
+**Accepted values**: `Cumulative`, `Delta`, `LowMemory`. <br>
+**Default**: `delta` <br>
+**Notes**: This default value is Datadog's recommended configuration and differs from the OpenTelemetry specification's default.
```
We should link to our documentation for why it's the Datadog-recommended configuration: https://docs.datadoghq.com/opentelemetry/guide/otlp_delta_temporality/

```suggestion
**Notes**: This default value `delta` is [Datadog's recommended configuration][Link#] and differs from the OpenTelemetry specification's default.
```

zacharycmontoya commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -29,21 +29,36 @@ further_reading:
-: ****Datadog convention****: `DD_SERVICE`<br>
-Sets the service name<br>
-**Notes**: If `service.name` is also provided in `OTEL_RESOURCE_ATTRIBUTES`, then `OTEL_SERVICE_NAME` takes precedence<br>
+`service.name` (Resource Attribute)
```
I think we should revert to the previous description while also adding the `service.name` resource attribute info to the `OTEL_RESOURCE_ATTRIBUTES` description later.

> This is resolved as a resource attribute, not from `OTEL_SERVICE_NAME`.
I'm not sure where this info came from, but `OTEL_SERVICE_NAME` [is responsible](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration) for setting the `service.name` resource attribute. It's a little weird because in our traces we don't really use resource attributes, but in our metrics and logs we do.

zacharycmontoya commented on content/en/opentelemetry/config/environment_variable_support.md:
In my mental model, I think of the hierarchy of technologies (and as a result, the documentation pages) as the following:
- Using OTel SDK
- Using DD SDK
  - Configuration
  - APIs
    - Lang-specific pages
  - Instrumentation Libraries
    - Lang-specific pages

What do you think of this hierarchy? And is it possible for us to implement this hierarchy in the docs structure or is the structure too nested?

brett0000FF commented on content/en/opentelemetry/instrument/api_support/python/metrics.md:
````diff
@@ -0,0 +1,110 @@
+   ```
+4. Instrument your application:
+   ```py
+   [Code example]
````
Thanks! Just attempted to add some examples to the Python Metrics page, if you could please confirm/adjust as needed?

brett0000FF commented:
## Open questions
- Minimum SDK Versions: What is the minimum required version of the Datadog SDK for this feature?
  - dd-trace-py: [x.y.z] or later?
  - dd-trace-dotnet: [x.y.z] or later?
  - dd-trace-js: [x.y.z] or later?
- Python: Please verify all code examples, especially the histogram example for Python Metrics. @mabdinur 
- .NET: Please verify and fill in all code examples. @link04 
- Node.js: Please verify and fill in all code examples.
- Environment Variable Support: I changed the default for`OTEL_EXPORTER_OTLP_LOGS_PROTOCOL` to `http/protobuf` but it seems like there is conflicting info the RFCs.
- @duncanhewett : 
  - Let me know if you have any ideas to adjust the "why" or the story to include in the Overview sections.
  - Also, `To avoid confusion I think it would be good to also have the Metrics/Logs tabs on the other languages and say 'coming soon' with a link to contact support for more details.` I would advocate that we _don't_ add the placeholder tabs right now. It's generally best practice to document what is supported at present. Currently, a user would click on a language we don't support yet and just see the traces API content, which I think is cleaner. Instead, how about we add a compatibility table to our Compatibility page? Just added in the latest commit. Let me know what you think!

And generally, please continue to verify the technical accuracy of all the new content. Thank you!
CC @zacharycmontoya

duncanhewett commented:
>   * Also, `To avoid confusion I think it would be good to also have the Metrics/Logs tabs on the other languages and say 'coming soon' with a link to contact support for more details.` I would advocate that we _don't_ add the placeholder tabs right now. It's generally best practice to document what is supported at present. Currently, a user would click on a language we don't support yet and just see the traces API content, which I think is cleaner. Instead, how about we add a compatibility table to our Compatibility page? Just added in the latest commit. Let me know what you think!
> 
That's fine with me - the compatibility page looks good!

git-thuerk-done commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -71,38 +93,155 @@ Mapped values between `OTEL_TRACES_SAMPLER` & `DD_TRACE_SAMPLE_RATE`:<br>
 `OTEL_INSTRUMENTATION_COMMON_DEFAULT_ENABLED`
-: ****Datadog convention****: `!DD_INTEGRATIONS_ENABLED` <br>
+: **Datadog convention**: `!DD_INTEGRATIONS_ENABLED` <br>
 Set to `false` to disable all instrumentation in the agent<br>
```
```suggestion
Set to `false` to disable all instrumentation in the Agent<br>
```
noticed this was not capitalized before but just correcting it here, unless this refers to an OTEL agent ?

git-thuerk-done commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,90 @@
+
+This feature works by intercepting logs from the built-in `Microsoft.Extensions.Logging.ILogger` library and exporting them as structured OTLP data.
+
+<div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature</div>
```
```suggestion
<div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature.</div>
```

git-thuerk-done commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,90 @@
+
+### Existing OTel setup (manual configuration)
+
+This Datadog feature is not compatible with applications already using the OpenTelemetry Logs API or OpenTelemetry SDK for logs.
```
```suggestion
This Datadog feature is not compatible with applications already using the OpenTelemetry Logs API or OpenTelemetry SDK for logs.
```
This seems like a fairly important callout, like if you are using X or Y, do not try these steps, may be worth converting to a Note callout ?

git-thuerk-done commented on content/en/opentelemetry/instrument/api_support/python/logs.md:
```diff
@@ -0,0 +1,134 @@
+
+All OTLP exporter settings (such as endpoints, protocols, and timeouts), resource attributes, and batch processor settings are configured using a shared set of OpenTelemetry environment variables.
+
+For a complete list of all shared OTLP environment variables, see [OpenTelemetry Environment Variables Interoperability][1].
```
```suggestion
For a complete list of all shared OTLP environment variables, see [OpenTelemetry Environment Variables Interoperability][1].
```
same with this link :)

git-thuerk-done replied:
I think it should link to [2]:

git-thuerk-done commented on content/en/opentelemetry/instrument/api_support/python/logs.md:
```diff
@@ -0,0 +1,134 @@
+
+1.  Set the `DD_LOGS_OTEL_ENABLED=true` environment variable.
+2.  The Datadog SDK automatically disables the old log injection style (`DD_LOGS_INJECTION`) to prevent duplicate trace metadata in your logs. Trace correlation is handled by the structured OTLP payload.
+3.  Ensure your Datadog Agent is configured to receive OTLP logs (version 7.48.0 or greater is required) and disable any file-based log collection for this service to avoid duplicate logs.
```
```suggestion
3.  Ensure your Datadog Agent is configured to receive OTLP logs (version 7.48.0 or greater is required) and disable any file-based log collection for this service to avoid duplicate logs.
```
same comment here, consider breaking to 2 steps

brett0000FF commented on content/en/opentelemetry/config/environment_variable_support.md:
```diff
@@ -71,38 +93,155 @@ Mapped values between `OTEL_TRACES_SAMPLER` & `DD_TRACE_SAMPLE_RATE`:<br>
+
+## OpenTelemetry Metrics SDK configuration
+
+Datadog SDKs support the following OpenTelemetry Metrics SDK options.
```
It uses an apostrophe when it is the possessive form. For example "Enables the Datadog SDK's OpenTelemetry interoperability for traces." uses the apostrophe because it is referring to the SDK's interoperability for traces. In this case, it is just plural, so we don't need an apostrophe.

mabdinur commented on content/en/opentelemetry/instrument/api_support/nodejs/metrics.md:
```diff
@@ -0,0 +1,114 @@
+
+## Prerequisites
+
+- **Datadog SDK**: dd-trace version [x.y.z] or later.
```
(will be released in 2 weeks)

```suggestion
- **Datadog SDK**: dd-trace version [5.77.0] or later.
```

mabdinur commented on content/en/opentelemetry/instrument/api_support/nodejs/metrics.md:
```diff
@@ -0,0 +1,114 @@
+## Prerequisites
+
+- **Datadog SDK**: dd-trace version [x.y.z] or later.
+- **OpenTelemetry API**: `@opentelemetry/api` version [x.y.z] or later. (The Datadog SDK provides the implementation for this API).
```
```suggestion
- **OpenTelemetry API**: `@opentelemetry/api` version 1.0.0 to 1.10.0. (The Datadog SDK provides the implementation for this API).
```

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
```diff
@@ -0,0 +1,92 @@
+
+## Prerequisites
+
+- **Datadog SDK**: dd-trace-dotnet version [x.y.z] or later.
```
Our support only runs on applications running on .NET 6 or newer. Maybe we add a line something like this?
```suggestion
- The instrumented application is running on .NET 6 or later.
- **Datadog SDK**: dd-trace-dotnet version [x.y.z] or later.
```

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
````diff
@@ -0,0 +1,92 @@
+
+1. Install the Datadog SDK:
+   ```sh
+   [TODO: Code example]
````
Installing the .NET Tracer is not a one-line instruction like other languages. Can we link to https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/dd_libraries/dotnet-framework/?tab=nuget#install-the-tracer for .NET Framework and https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/dd_libraries/dotnet-core?tab=windows#install-the-tracer for .NET Core?

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
````diff
@@ -0,0 +1,92 @@
+This example uses the OTel Metrics API to create a counter that increments every time an item is processed:
+
+```csharp
+[TODO: Code example]
````
```suggestion
using System.Diagnostics.Metrics;

// Define a meter
Meter meter = new("MyService", "1.0.0");

// Create a counter instrument, which will be used to record measurements in your code
Counter<long> requestsCounter = meter.CreateCounter<long>("http.requests_total");

// Perform work
// ...

// Record measurements
requestsCounter.Add(1, new("method", "GET"), new("status_code", "200"));
```

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
````diff
@@ -0,0 +1,92 @@
+This example uses the OTel Metrics API to create a histogram to track request durations:
+
+```csharp
+[TODO: Code example]
````
```suggestion
using System.Diagnostics.Metrics;

// Define a meter
Meter meter = new("MyService", "1.0.0");

// Create a histogram instrument, which will be used to record measurements in your code
Histogram<double> responseTimeHistogram = meter.CreateHistogram<double>("http.response.time");

// Perform work
var watch = System.Diagnostics.Stopwatch.StartNew();
await Task.Delay(1_000);
watch.Stop();

// Record measurements
responseTimeHistogram.Record(watch.ElapsedMilliseconds, new("method", "GET"), new("status_code", "200"));
```

brett0000FF commented on content/en/opentelemetry/instrument/api_support/nodejs/logs.md:
```diff
@@ -0,0 +1,142 @@
+
+## Prerequisites
+
+- **Datadog SDK**: `dd-trace-js` version [x.y.z] or later.
```
Thanks for the info! I added to the draft here: https://github.com/DataDog/documentation/pull/32610/commits/bd53c9a2b88f11725052007f79bc3a2401ca45bf.

Also, I forgot this was still in alpha, and [we omitted it from the feature compatibility table](https://docs-staging.datadoghq.com/brett.blue/otel-log-metric-phase-1/opentelemetry/compatibility/#api-support). I hide the page for now. When it is ready, we can remove the private tag in the frontmatter and uncomment out the mutli-code-lang params.

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
```diff
@@ -0,0 +1,120 @@
+If you are already using the OpenTelemetry SDK with a manual OTLP exporter configuration, follow these steps to migrate:
+
+1. Add the Datadog SDK (`dd-trace-dotnet`) to your project and enable its instrumentation.
+2. Remove any code that manually configures the `OTLPMetricsExporter`. The Datadog SDK handles this configuration automatically.
```
```suggestion
2. Remove any code that manually configures the `OtlpExporter` for metrics. The Datadog SDK handles this configuration automatically.
```

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,92 @@
+## Troubleshooting
+
+{{% otel-api-troubleshooting signal="logs" %}}
+- Verify the Datadog Profiler is attached. This feature relies on Datadog's automatic instrumentation. Ensure the .NET Profiler is correctly enabled for your application.
```
Same comment as https://github.com/DataDog/documentation/pull/32610/files#r2516341501 . "Datadog Profiler" can be confusing. Can we reword so it's not confused with the Datadog Continuous Profiler product?

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,92 @@
+
+This feature works by intercepting logs from the built-in `Microsoft.Extensions.Logging.ILogger` library and exporting them as structured OTLP data.
+
+<div class="alert alert-info">This implementation does not use the OpenTelemetry Logs API. It is not compatible with the <code>OpenTelemetry.SDK.Logs</code> or <code>OpenTelemetry.Exporter.Otlp.Logs</code> packages. You must remove them to use this feature.</div>
```
I think the "Native" callout is also sufficient here
```suggestion
<div class="alert alert-info">You should not install the official OpenTelemetry SDK or any OTLP Exporter packages. The Datadog SDK provides this functionality. Installing both can lead to runtime conflicts and duplicate data.</div>
```

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
```diff
@@ -0,0 +1,92 @@
+
+## Troubleshooting
+
+{{% otel-api-troubleshooting signal="logs" %}}
```
I wonder if we could do a list "start" and "end" snippet so we can add additional items into the same `<ul>` HTML section. WDYT?

brett0000FF commented on content/en/opentelemetry/instrument/api_support/dotnet/metrics.md:
```diff
@@ -0,0 +1,142 @@
+  - `ObservableGauge`
+
+- **.NET 7+** (or `System.Diagnostics.DiagnosticSource` v7.0.0):
+  - *Adds support for:*
```
I like it!

zacharycmontoya commented on content/en/opentelemetry/instrument/api_support/dotnet/logs.md:
````diff
@@ -0,0 +1,90 @@
+[TODO: Code example]
+```
+
+### Trace and log correlation
````
I think we can also link to the traces page for the "OTel Tracing API support" text.

```suggestion
### Trace and log correlation

This example shows how logs emitted within an active Datadog span are automatically correlated. If you are using the OTel Tracing API or built-in .NET Activity API to create spans, ensure OTel Tracing API support is enabled.
```

mabdinur commented on content/en/opentelemetry/instrument/api_support/nodejs/metrics.md:
```diff
@@ -0,0 +1,116 @@
+
+## Prerequisites
+
+- **Datadog SDK**: dd-trace version 5.77.0 or later.
```
```suggestion
- **Datadog SDK**: dd-trace version 5.60.0 or later.
```
